pyroscope/symb/irsymcache/cache.go

```diff
@@ -0,0 +1,381 @@
+		errored:  make(map[libpf.FileID]struct{}),
+		enabled:  opt.Enabled,
+	}
+	res.cacheDir = filepath.Join(res.cacheDir, impl.Name())
+	err := os.MkdirAll(res.cacheDir, 0o700)
+	if err != nil {
+		return nil, err
+	}
+	lru.SetOnEvict(func(id libpf.FileID, sz int) {
+		filePath := res.tableFilePath(id)
+		l.WithFields(logrus.Fields{
+			"file": filePath,
+			"size": sz,
+		}).Debug("symbcache evicting")
+		_ = os.Remove(filePath)
+		res.mutex.Lock()
+		delete(res.known, id)
+		res.mutex.Unlock()
+	})
+
+	// list dir and add to cache
+	err = filepath.Walk(res.cacheDir, func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+		if info.IsDir() {
+			return nil
+		}
+		id, err := libpf.FileIDFromString(filepath.Base(path))
+		if err != nil {
+			return nil
+		}
+		id2 := id.StringNoQuotes()
+		if filepath.Base(path) != id2 {
+			return nil
+		}
+
+		res.lru.Put(id, int(info.Size()))
+		res.known[id] = struct{}{}
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	// Start monitoring goroutine
+	res.wg.Add(1)
+	go func() {
+		defer res.wg.Done()
+		i := 0
+		ticker := time.NewTicker(time.Minute)
+		defer ticker.Stop()
+
+		for {
+			select {
+			case <-shutdown:
+				return
+			case <-ticker.C:
+				l.WithField("size", lru.Size()).Info("fscache lru size")
+				i++
+				if i%10 == 0 {
+					res.mutex.Lock()
+					res.errored = make(map[libpf.FileID]struct{})
+					res.mutex.Unlock()
+				}
+			}
+		}
+	}()
+
+	res.wg.Add(1)
+	go func() {
+		defer res.wg.Done()
+		convertLoop(res)
+	}()
+
+	return res, nil
+}
+
+func convertLoop(res *Resolver) {
+	runtime.LockOSThread()
+	defer runtime.UnlockOSThread()
+
+	for {
+		select {
+		case <-res.shutdown:
+			for len(res.jobs) > 0 {
+				job := <-res.jobs
+				job.result <- res.convertSync(job.src, job.dst)
+			}
+			return
+		case job := <-res.jobs:
+			job.result <- res.convertSync(job.src, job.dst)
+		}
+	}
+}
+
+func (c *Resolver) Observe(fid libpf.FileID, elfRef *pfelf.Reference) {
+	o, ok := elfRef.ELFOpener.(pfelf.RootFSOpener)
+	if !ok {
+		return
+	}
+	if !c.enabled {
+		return
+	}
+	if elfRef.FileName() == process.VdsoPathName {
+		return
+	}
+
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	if c.isKnown(fid) {
+		return
+	}
+	pid := 0
+	if pp, ok := elfRef.ELFOpener.(process.Process); ok {
+		pid = int(pp.PID())
+	}
+	l := c.logger.WithFields(logrus.Fields{
+		"fid": fid.StringNoQuotes(),
+		"elf": elfRef.FileName(),
+		"pid": pid,
+	})
+	t1 := time.Now()
+	err := c.convert(l, fid, elfRef, o)
+	if err != nil {
+		l.WithError(err).Error("conversion failed")
+		c.errored[fid] = struct{}{}
+	} else {
+		l.Debug("converted")
+		c.known[fid] = struct{}{}
+	}
+	l.WithField("duration", time.Since(t1)).Debug("conversion completed")
+}
+
+func (c *Resolver) convert(
+	l *logrus.Entry,
+	fid libpf.FileID,
+	elfRef *pfelf.Reference,
+	o pfelf.RootFSOpener,
+) error {
+	var err error
+	var dst *os.File
+	var src *os.File
+
+	_, _ = c.lru.Get(fid)
```

Comment:
I must be missing something, but this is the only place we get from the LRU cache. I would expect we "touch" a cached item every time we lookup anything from the lookup table. Otherwise our cache works like a FIFO queue.